File: iosApp/iosApp/offer/validation/OpenOfferValidator.swift
//
//  OpenOfferValidator.swift
//  iosApp
//
//  Created by jimmyt on 1/6/23.
//  Copyright © 2023 orgName. All rights reserved.
//

/**
 Ensures that `offer` can be opened.
 
 This function ensures that:
    - the user of this interface is the maker of `offer`.
    - `offer` is in the `OfferState.awaitingOpening` state.
    - `offer` is not already being opened.
 
 - Parameter offer: The `Offer` to be validated for opening.

 - Throws: An `OfferDataValidationError` if this is not able to ensure any of the conditions in the list above. The descriptions of the errors thrown by this function are human-readable and can be displayed to the user so that they can correct any problems.
 */
func validateOfferForOpening(offer: Offer) throws {
    guard offer.isUserMaker else {
        throw OfferDataValidationError(desc: "You can only open your own offers")
    }
    guard offer.state == .awaitingOpening else {
        throw OfferDataValidationError(desc: "This Offer cannot currently be opened.")
    }
    guard offer.openingOfferState == .none || offer.openingOfferState == .validating || offer.openingOfferState == .error else {
        throw OfferDataValidationError(desc: "This Offer is already being opened.")
    }
}

File: androidApp/src/main/java/com/commuto/interfacemobile/android/offer/validation/OpenOfferValidator.kt
package com.commuto.interfacemobile.android.offer.validation

import com.commuto.interfacemobile.android.offer.Offer
import com.commuto.interfacemobile.android.offer.OfferState
import com.commuto.interfacemobile.android.offer.OpeningOfferState

/**
 * Ensures that [offer] can be opened.
 *
 * This function ensures that:
 * - the user of this interface is the maker of [offer].
 * - [offer] is in the [OfferState.AWAITING_OPENING] state.
 * - [offer] is not already being opened.
 *
 * @param offer The [Offer] to be validated for opening.
 *
 * @throws [OfferDataValidationException] if this is not able to ensure any of the conditions in the list above. The
 * descriptions of the exceptions thrown by this function are human-readable and can be displayed to the user so that
 * they can correct any problems.
 */
fun validateOfferForOpening(offer: Offer) {
    if (!offer.isUserMaker) {
        throw OfferDataValidationException(desc = "You can only open your own offers")
    }
    if (offer.state != OfferState.AWAITING_OPENING) {
        throw OfferDataValidationException(desc = "This Offer cannot currently be opened.")
    }
    if (offer.openingOfferState.value != OpeningOfferState.NONE &&
        offer.openingOfferState.value != OpeningOfferState.VALIDATING &&
        offer.openingOfferState.value != OpeningOfferState.EXCEPTION) {
        throw OfferDataValidationException(desc = "This Offer is already being opened.")
    }
}